src/js/tracing/stalltracking.ts

```diff
@@ -76,100 +80,76 @@ export class StallTrackingInstrumentation {
-      const originalAdd = transaction.spanRecorder.add;
-
-      transaction.spanRecorder.add = (span: Span): void => {
-        originalAdd.apply(transaction.spanRecorder, [span]);
-
-        // eslint-disable-next-line @typescript-eslint/unbound-method
-        const originalSpanFinish = span.finish;
-
-        span.finish = (endTimestamp?: number) => {
-          // We let the span determine its own end timestamp as well in case anything gets changed upstream
-          originalSpanFinish.apply(span, [endTimestamp]);
-
-          // The span should set a timestamp, so this would be defined.
-          if (span.endTimestamp) {
-            this._markSpanFinish(transaction, span.endTimestamp);
-          }
-        };
-
-        // eslint-disable-next-line @typescript-eslint/unbound-method
-        const originalSpanEnd = span.end;
-
-        span.end = (endTimestamp?: number) => {
-          // We let the span determine its own end timestamp as well in case anything gets changed upstream
-          originalSpanEnd.apply(span, [endTimestamp]);
-
-          // The span should set a timestamp, so this would be defined.
-          if (span.endTimestamp) {
-            this._markSpanFinish(transaction, span.endTimestamp);
-          }
-        };
-      };
-    }
-  }
+  };
 
   /**
    * Logs a transaction as finished.
    * Stops stall tracking if no more transactions are running.
    * @returns The stall measurements
    */
-  public onTransactionFinish(transaction: Transaction | IdleTransaction, passedEndTimestamp?: number): void {
-    const transactionStats = this._statsByTransaction.get(transaction);
+  private _onSpanEnd = (rootSpan: Span, passedEndTimestamp?: number): void => {
+    if (!isRootSpan(rootSpan)) {
+      return this._onChildSpanEnd(rootSpan);
+    }
+
+    const transactionStats = this._statsByRootSpan.get(rootSpan);
 
     if (!transactionStats) {
       // Transaction has been flushed out somehow, we return null.
       logger.log('[StallTracking] Stall measurements were not added to transaction due to exceeding the max count.');
 
-      this._statsByTransaction.delete(transaction);
+      this._statsByRootSpan.delete(rootSpan);
       this._shouldStopTracking();
 
       return;
     }
 
-    const endTimestamp = passedEndTimestamp ?? transaction.endTimestamp;
+    const endTimestamp = passedEndTimestamp ?? spanToJSON(rootSpan).timestamp;
 
-    const spans = transaction.spanRecorder ? transaction.spanRecorder.spans : [];
-    const finishedSpanCount = spans.reduce((count, s) => (s !== transaction && s.endTimestamp ? count + 1 : count), 0);
+    const spans = getSpanDescendants(rootSpan);
+    const finishedSpanCount = spans.reduce(
+      (count, s) => (s !== rootSpan && spanToJSON(s).timestamp ? count + 1 : count),
+      0,
+    );
 
-    const trimEnd = transaction.toContext().trimEnd;
+    // TODO: Transaction will be removed, can we replace trimEnd with lastSpan.end_timestamp === rootSpan.end_timestamp?
```

Comment:
We no longer have transactions, should this be addressed before merging this PR?

Reply:
Thank you this is good catch, I've updated the tests and the impl.